libfreerdp/core/redirection.c

```diff
@@ -481,27 +593,354 @@ void redirection_free(rdpRedirection* redirection)
+		Stream_Write_UINT32(s, redirection->sessionID);
+	else
+		Stream_Write_UINT32(s, 0);
+
+	Stream_Write_UINT32(s, redirection->flags);
+
+	if (redirection->flags & LB_TARGET_NET_ADDRESS)
+	{
+		if (redir_write_string(LB_TARGET_NET_ADDRESS, s, redirection->TargetNetAddress) < 0)
+			goto fail;
+	}
+
+	if (redirection->flags & LB_LOAD_BALANCE_INFO)
+	{
+		const UINT32 length = 13 + redirection->LoadBalanceInfoLength + 2;
+		if (!Stream_EnsureRemainingCapacity(s, length))
+			goto fail;
+		Stream_Write_UINT32(s, length);
+		Stream_Write(s, "Cookie: msts=", 13);
+		Stream_Write(s, redirection->LoadBalanceInfo, redirection->LoadBalanceInfoLength);
+		Stream_Write_UINT8(s, 0x0d);
+		Stream_Write_UINT8(s, 0x0a);
+	}
+
+	if (redirection->flags & LB_USERNAME)
+	{
+		if (redir_write_string(LB_USERNAME, s, redirection->Username) < 0)
+			goto fail;
+	}
+
+	if (redirection->flags & LB_DOMAIN)
+	{
+		if (redir_write_string(LB_DOMAIN, s, redirection->Domain) < 0)
+			goto fail;
+	}
+
+	if (redirection->flags & LB_PASSWORD)
+	{
+		/* Password is eighter UNICODE or opaque data */
+		if (!redir_write_data(LB_PASSWORD, s, redirection->PasswordLength, redirection->Password))
+			goto fail;
+	}
+
+	if (redirection->flags & LB_TARGET_FQDN)
+	{
+		if (redir_write_string(LB_TARGET_FQDN, s, redirection->TargetFQDN) < 0)
+			goto fail;
+	}
+
+	if (redirection->flags & LB_TARGET_NETBIOS_NAME)
+	{
+		if (redir_write_string(LB_TARGET_NETBIOS_NAME, s, redirection->TargetNetBiosName) < 0)
+			goto fail;
+	}
+
+	if (redirection->flags & LB_CLIENT_TSV_URL)
+	{
+		if (!redir_write_data(LB_CLIENT_TSV_URL, s, redirection->TsvUrlLength, redirection->TsvUrl))
+			goto fail;
+	}
+
+	if (redirection->flags & LB_REDIRECTION_GUID)
+	{
+		if (!redir_write_data(LB_REDIRECTION_GUID, s, redirection->RedirectionGuidLength,
+		                      redirection->RedirectionGuid))
+			goto fail;
+	}
+
+	if (redirection->flags & LB_TARGET_CERTIFICATE)
+	{
+		if (!redir_write_data(LB_REDIRECTION_GUID, s, redirection->TargetCertificateLength,
+		                      redirection->TargetCertificate))
+			goto fail;
+	}
+
+	if (redirection->flags & LB_TARGET_NET_ADDRESSES)
+	{
+		UINT32 i;
+		UINT32 length = sizeof(UINT32);
+
+		if (!Stream_EnsureRemainingCapacity(s, 2 * sizeof(UINT32)))
+			goto fail;
+
+		const size_t start = Stream_GetPosition(s);
+		Stream_Seek_UINT32(s); /* length of field */
+		Stream_Write_UINT32(s, redirection->TargetNetAddressesCount);
+		for (i = 0; i < redirection->TargetNetAddressesCount; i++)
 		{
-			int i;
+			const SSIZE_T rcc =
+			    redir_write_string(LB_TARGET_NET_ADDRESSES, s, redirection->TargetNetAddresses[i]);
+			if (rcc < 0)
+				goto fail;
+			length += (UINT32)rcc;
+		}
+
+		/* Write length field */
+		const size_t end = Stream_GetPosition(s);
+		Stream_SetPosition(s, start);
+		Stream_Write_UINT32(s, length);
+		Stream_SetPosition(s, end);
+	}
+
+	if (!Stream_EnsureRemainingCapacity(s, 1))
+		goto fail;
+	Stream_Write_UINT8(s, 0);
```

Comment:
I assume you mean the `Pad` here? Should these be 8Bytes instead of 8Bits?

```
Pad (8 bytes): An optional 8-element array of 8-bit unsigned integers. Padding. Values in this field
MUST be ignored.
```

Reply:
updated